dcm4che-imageio/src/main/java/org/dcm4che3/imageio/plugins/dcm/ParseEntireFileMetaData.java

```diff
@@ -0,0 +1,191 @@
+//        Agfa HealthCare N.V. and/or its affiliates.        //
+//      The copyright notice above does not evidence any     //
+//     actual or intended publication of such source code.   //
+//                                                           //
+///////////////////////////////////////////////////////////////
+//
+package org.dcm4che3.imageio.plugins.dcm;
+
+import org.dcm4che3.data.*;
+import org.dcm4che3.error.NoPixelDataException;
+import org.dcm4che3.image.PhotometricInterpretation;
+import org.dcm4che3.imageio.stream.BulkURIImageInputStream;
+import org.dcm4che3.imageio.stream.ByteArrayImageInputStream;
+import org.dcm4che3.io.stream.ImageInputStreamLoader;
+import org.dcm4che3.io.stream.ServiceImageInputStreamLoader;
+import org.dcm4che3.io.stream.DelegateImageInputStreamLoader;
+import org.dcm4che3.imageio.stream.ImageInputStreamAdapter;
+import org.dcm4che3.imageio.stream.SegmentedImageInputStream;
+import org.dcm4che3.io.BulkDataDescriptor;
+import org.dcm4che3.io.DicomInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.imageio.stream.ImageInputStream;
+import javax.imageio.stream.MemoryCacheImageInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.ByteOrder;
+import java.nio.file.Path;
+
+/**
+ * Load the entire file into memory as BulkData.  Pixels are streamed from the BulkData URI to clients.  This class
+ * sacrifices initial parse time for having a stateless class which is able to access the same file from different threads
+ * without synchronization.
+ *
+ */
+public class ParseEntireFileMetaData extends DicomMetaData {
+    private static final Logger LOG = LoggerFactory.getLogger(ParseEntireFileMetaData.class);
+
+    private static final ImageInputStreamLoader<Object> SERVICE_LOADER = new ServiceImageInputStreamLoader();
+
+    private final Attributes fmi;
+    private final Attributes attributes;
+    private final URI uri;
+    private final ImageInputStreamLoader<URI> uriLoader;
+
+    private final Object pixelData;
+    private final VR pixelDataVR;
+
+    // IIS that was used to load this file.
+
+
+    public ParseEntireFileMetaData(Object input) throws IOException {
+        this.uri = toURI(input);
+
+
+        String uriStr = this.uri.toString();
+        ImageInputStream iis = SERVICE_LOADER.openStream(input);
+        try (DicomInputStream dis = new DicomInputStream(new ImageInputStreamAdapter(iis))) {
+            dis.setIncludeBulkData(DicomInputStream.IncludeBulkData.URI);
+            dis.setBulkDataDescriptor(BulkDataDescriptor.PIXELDATA);
+            dis.setURI(uriStr);
+
+            this.fmi = dis.getFileMetaInformation();
+            this.attributes = dis.readDataset(-1,-1);
+            this.pixelDataVR = attributes.getVR(Tag.PixelData);
+            this.pixelData = attributes.remove(Tag.PixelData);
```

Comment:
I very much like the fact that by the time this method is done, the pixelData is fully defined, although I suspect it will be necessary to have various lazy versions of this that use a separate thread to finish filling out the data.

Reply:
Yes, I think so too.  For the default implementation  I wanted something very simple.  The DicomImageAccessor can accommodate more complex designs, as the openPixelStream(int frameIndex) call hides the particulars.